Add destroy to BlockTemplate schema

Sjors opened the pull request:
This ensures that if a client no longer needs a block template, the node can clear its memory as soon as possible.

A block template may hold on to transactions that are no longer in the mempool, so this can be significant.

This has a trivial silent merge conflict with #31283 because it also used the number `@9` (gaps are not allowed). I'll rebase whichever is merged last.

Sjors commented:
@ryanofsky can you further clarify what the difference in behavior is with an without a `destroy` method?

ryanofsky commented on src/ipc/capnp/mining.capnp:
```diff
@@ -33,6 +33,7 @@ interface BlockTemplate $Proxy.wrap("interfaces::BlockTemplate") {
     getWitnessCommitmentIndex @6 (context: Proxy.Context) -> (result: Int32);
     getCoinbaseMerklePath @7 (context: Proxy.Context) -> (result: List(Data));
     submitSolution@8 (context: Proxy.Context, version: UInt32, timestamp: UInt32, nonce: UInt32, coinbase :Data) -> (result: Bool);
+    destroy @9 (context :Proxy.Context) -> ();
```
In commit "Add destroy to BlockTemplate schema" (5f7dc0dbe64eb38afc74e9c4cba489382ea373bd)

Would suggest making destroy the first method because it affect semantics of the object as a whole, and thats where I placed destroy in all the other interfaces. It is fine to keep the number @9 for it, or renumber if you prefer.

Sjors replied:
I moved `destroy` to the top and renumbered.

ryanofsky reviewed (approved):
Code review ACK 5f7dc0dbe64eb38afc74e9c4cba489382ea373bd

> @ryanofsky can you further clarify what the difference in behavior is with an without a `destroy` method?

I think it probably doesn't matter too much in this case, since the only thing the destroy method will currently do is free memory. But this does help future proof the interface, in case we want the destructor to do more things in the future, or if we add memory limits and want to give clients a way to ensure that previous block templates were completely freed before they create new ones.

The difference technically between having a destroy method and not having one is that when a libmultiprocess c++ IPC client destroys an object, it will automatically call the destroy method and wait for the server object to be destroyed before returning to client code. Without a destroy method, the C++ client object will be destroyed immediately without waiting for the server object to be destroyed. There's some documentation about this in https://github.com/chaincodelabs/libmultiprocess/blob/abe254b9734f2e2b220d1456de195532d6e6ac1e/include/mp/proxy.h#L80-L91

sedited reviewed (approved):
ACK b28972cd85e4472b386349d6cda8c233faeffd4f

ryanofsky commented on src/ipc/capnp/mining.capnp:
```diff
@@ -24,15 +24,16 @@ interface Mining $Proxy.wrap("interfaces::Mining") {
+    getCoinbaseCommitment @6 (context: Proxy.Context) -> (result: Data);
+    getWitnessCommitmentIndex @7 (context: Proxy.Context) -> (result: Int32);
+    getCoinbaseMerklePath @8 (context: Proxy.Context) -> (result: List(Data));
+    submitSolution@9 (context: Proxy.Context, version: UInt32, timestamp: UInt32, nonce: UInt32, coinbase :Data) -> (result: Bool);
```
In commit "Add destroy to BlockTemplate schema" (b28972cd85e4472b386349d6cda8c233faeffd4f)

note: Missing a space before the number this line

Sjors replied:
Added the space.

ryanofsky reviewed (approved):
Code review ACK b28972cd85e4472b386349d6cda8c233faeffd4f

Sjors committed:
Add destroy to BlockTemplate schema

This ensures that if a client no longer needs a block template,
the node can clear its memory as soon as possible.

A block template may hold on to transactions that are no longer
in the mempool, so this can be significant.

sedited reviewed (approved):
Re-ACK 9aa50152c1cfa1c41215b2b51ed7a516aa67137a

ryanofsky reviewed (approved):
Code review ACK 9aa50152c1cfa1c41215b2b51ed7a516aa67137a

ryanofsky commented:
Will go ahead and merge even though it only has 2 acks, since it's basically just adding a line to a schema file. Having this merged should make the related PR easier to maintain and review.
